pkg/cloudprovider/provider/packet/provider.go

```diff
@@ -0,0 +1,459 @@
+		c.BillingCycle = defaultBillingCycle
+	}
+	for i, tag := range rawConfig.Tags {
+		tagValue, err := p.configVarResolver.GetConfigVarStringValue(tag)
+		if err != nil {
+			return nil, nil, nil, fmt.Errorf("failed to get the %d value of \"tags\" field, error = %v", i, err)
+		}
+		c.Tags = append(c.Tags, tagValue)
+	}
+	for i, facility := range rawConfig.Facilities {
+		facilityValue, err := p.configVarResolver.GetConfigVarStringValue(facility)
+		if err != nil {
+			return nil, nil, nil, fmt.Errorf("failed to get the %d value of \"facilities\" field, error = %v", i, err)
+		}
+		c.Facilities = append(c.Facilities, facilityValue)
+	}
+	return &c, &rawConfig, &pconfig, err
+}
+
+func (p *provider) getPacketDevice(machine *v1alpha1.Machine) (*packngo.Device, *packngo.Client, error) {
+	c, _, _, err := p.getConfig(machine.Spec.ProviderSpec)
+	if err != nil {
+		return nil, nil, cloudprovidererrors.TerminalError{
+			Reason:  common.InvalidConfigurationMachineError,
+			Message: fmt.Sprintf("Failed to parse MachineSpec, due to %v", err),
+		}
+	}
+
+	client := getClient(c.APIKey)
+	device, err := getDeviceByTag(client, c.ProjectID, generateTag(string(machine.UID)))
+	if err != nil {
+		return nil, nil, err
+	}
+	return device, client, nil
+}
+
+func (p *provider) Validate(spec v1alpha1.MachineSpec) error {
+	c, _, pc, err := p.getConfig(spec.ProviderSpec)
+	if err != nil {
+		return fmt.Errorf("failed to parse config: %v", err)
+	}
+
+	if c.APIKey == "" {
+		return errors.New("apiKey is missing")
+	}
+	if c.InstanceType == "" {
+		return errors.New("instanceType is missing")
+	}
+	if c.ProjectID == "" {
+		return errors.New("projectID is missing")
+	}
+
+	_, err = getNameForOS(pc.OperatingSystem)
+	if err != nil {
+		return fmt.Errorf("invalid/not supported operating system specified %q: %v", pc.OperatingSystem, err)
+	}
+
+	client := getClient(c.APIKey)
+
+	if len(c.Facilities) == 0 || c.Facilities[0] == "" {
+		return fmt.Errorf("must have at least one non-blank facility")
+	}
+
+	// get all valid facilities
+	facilities, _, err := client.Facilities.List(nil)
+	if err != nil {
+		return fmt.Errorf("failed to list facilities: %v", err)
+	}
+	// ensure our requested facilities are in those facilities
+	if missingFacilities := itemsNotInList(facilityProp(facilities, "Code"), c.Facilities); len(missingFacilities) > 0 {
+		return fmt.Errorf("unknown facilities: %s", strings.Join(missingFacilities, ","))
+	}
+
+	// get all valid plans a.k.a. instance types
+	plans, _, err := client.Plans.List(nil)
+	if err != nil {
+		return fmt.Errorf("failed to list instance types / plans: %v", err)
+	}
+	// ensure our requested plan is in those plans
+	validPlanNames := planProp(plans, "Name")
+	if missingPlans := itemsNotInList(validPlanNames, []string{c.InstanceType}); len(missingPlans) > 0 {
+		return fmt.Errorf("unknown instance type / plan: %s, acceptable plans: %s", strings.Join(missingPlans, ","), strings.Join(validPlanNames, ","))
+	}
+
+	return nil
```

Comment:
Should we also validate the BillingCycle ?

Reply:
I am good for that. The default, obviously, works, as we put in a sane default, but good to validate. Will be updated in 5 mins.

Reply:
Ah, I remember now why I did it. There isn't an API to validate a billing cycle. You put it in the `Create()` request, and if it isn't valid, it rejects it.